report/3.Theory/ParserImplementation.tex

```diff
@@ -0,0 +1,96 @@
+One major reason for doing in line contextual analysis is the way functions are invoked in HCL. 
+The only way to know whether anything is a function call, is by analyzing the types. 
+This means that in order to create a proper abstract syntax tree the type checking must be done in line. 
+It could also potentially be done afterwards, but that would require quite a big amount of post processing, that would be a lot more complex, than just doing type checking within the parser. 
+This also means that the outputted abstract syntax tree is in fact already a decorated syntax tree. 
+For instance function calls already has information on its return type, even though the function may have multiple return types based on the overloading of the function.
+
+\subsection{Implementing a recursive decent parser}
+
+A LL(K) recursive decent parser will do look aheads until it is aware what branch should be parsed.
+This can be figured out from the first and follow sets, that can be produced from the CFG.
+Once it figures out which non-terminal to parse it will call a parse method that should parse that part.
+The parse method can then accept terminals or parse further non-terminals.
+It should be the responsibility of the accept method to advance the lexical token stream.
+
+Lets have a quick and concrete example from HCL. 
+
+The line to parse is "num x = 5" The token stream to parse will look as follows:
+
+\begin{itemize}
+	\item Number type
+	\item Identifier = x
+	\item Equals
+	\item Number Literal = 5
+	\item End of line
+\end{itemize}
+
+The parser will see that the first token is a number type. 
+Since only declarations have types in its first set, the parser now knows that it should parse a declaration.
+
+The EBNF reveals that a declaration is:
+\begin{itemize}
+	\item Non-terminal Type
+	\item Terminal Identifier
+	\item Optional
+	\subitem Terminal Equals
+	\subitem Non-terminal Expression
+\end{itemize}
+
+Based on this, the parse declaration is implemented as seen in listing \ref{lis:parseDeclaration}.
+Do note, that the final parse declaration method is a little more complex due to type inference and type checking.
+
+\begin{lstlisting}[language=java,label=lis:parseDeclaration,caption=A simplified version of the parse declaration method from the parser.]
+private fun parseDeclaration(): TreeNode.Command.Declaration {
+    val type = parseType()
+    val identifierToken = accept<Token.Identifier>()
+    val identifier = AstNode.Command.Expression.Value.Identifier(identifierToken.value)
+    val expression = if (current.token == Token.SpecialChar.Equals) {
+	    accept<Token.SpecialChar.Equals>()
+	    parseExpression()
+    } else null
+    return TreeNode.Command.Declaration(type, identifier, expression)
+}
+\end{lstlisting}
+
+As seen in the listing, the method will parse the type, accept the identifier, accept the equal sign, and at last it will parse the expression.
```

Comment:
Change to: "As seen in the above listing, .........", perhaps even include a reference (optional) (Mostly for clarity)